src/data_traits.rs

```diff
@@ -125,6 +135,12 @@ unsafe impl<A> DataRaw for PtrRepr<*const A> {
     private_impl!{}
 }
 
+unsafe impl<A> DataRawClone for PtrRepr<*const A> {
+    unsafe fn clone_with_ptr(&self, ptr: *mut Self::Elem) -> (Self, *mut Self::Elem) {
+        (*self, ptr)
+    }
+}
```

Comment:
In current `ndarray`, [`DataClone`](https://docs.rs/ndarray/0.12.0/ndarray/trait.DataClone.html) has a `Data` bound. If we want to be able to clone `ArrayPtr`/`ArrayPtrMut`, this is too restrictive given  only the [current implementation of `Clone` for `ArrayBase`](https://docs.rs/ndarray/0.12.0/ndarray/struct.ArrayBase.html#impl-Clone). Possible alternatives to the approach in this PR:

1. Don't add a `DataRawClone` trait; instead, implement `Clone` individually for `ArrayPtr` and `ArrayPtrMut`. That would be inconsistent and would prevent code with just a `DataClone` bound from being able to clone `ArrayPtr` and `ArrayPtrMut`.

   With a breaking change, though, we could get rid of `DataClone` entirely and just implement `Clone` individually for `ArcArray`, `Array`, `ArrayView`, `ArrayPtr`, and `ArrayPtrMut`. This actually seems pretty nice, but it does make writing generic code more difficult. (In user code, the bound to allow `array.clone()` would have to be `ArrayBase<S, D>: Clone` instead of `S: DataRawClone`.)

2. Add the `DataRawClone` trait and remove the `DataClone` trait. That would work but would be a breaking change.

Fwiw, `DataClone` does have a small amount of utility as the combination of `Data` and `DataRawClone`, but I agree it does seem like unnecessary complication.

Reply:
(2) I think a breaking change is ok. We have lots of them queued? We should get on to merging them as soon as 0.12.1 is released (and I can do the releasing if you want).

I'd hope it has very little impact on users, but maybe I underestimate how much these representation traits are used?

I'd use the name `DataClone` for the new trait.

Reply:
> I'd use the name `DataClone` for the new trait.

The only thing that I don't like about this is consistency of naming with the other data traits, especially `DataMut` vs. `DataRawMut`. I was naming traits without a `Data` bound `DataRaw*` and traits with a `Data` bound `Data*`.

Reply:
Good point. We still should use only one trait if we don't need 2, we could use the name DataRawClone and let DataClone be a deprecated reexport of it.

Reply:
DataClone could be a trait alias of DataRawClone + Data? That's a new Rust feature we can transition to.

Reply:
Woah, I had no idea trait aliases were a thing. They look very convenient.

I've added a commit deprecating `DataClone`. We can switch `DataClone` to a trait alias once they're available on stable.